infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -375,6 +420,25 @@ long getMaxCacheSize() {
         return config.getMaxCacheSize();
     }
 
+    @VisibleForTesting
+    BatchProcessor getBatchWriter() {
+        return batchWriter;
+    }
+
+    // The following methods should only be used for unit tests, ideally the executor should be
+    // final, but this "hack" is needed to now when a task as completed
+    @VisibleForTesting
+    void startHandler() {
```

Comment:
Java's Executor/ExecutorService don't have an API to track pending tasks without invoking shutdown or being the caller, so when a requests spans multiple executors to complete there's no way of figuring out when all the executors are "done", so by shutting down the executor and installing a new one, we implicitly wait on the tasks to complete.

Is there a better way to do this in java? 

When a request spans multiple executors

Reply:
ExecutorCompletionService has an api take() which returns a completed task. I'm not sure if this is useful in your case.